packages/ra-ui-materialui/src/list/ListView.tsx

```diff
@@ -130,42 +52,50 @@ export const ListView = props => {
         loading,
         hasCreate,
         filterValues,
-    } = rest;
-    const controllerProps = getListControllerProps(rest);
+        selectedIds,
+    } = listContext;
+
+    // @deprecated to be removed in 4.0
+    // we inject listContext several times, for backwards compatibility reasons
+    // children should use useListContext instead to grab the controller props
 
     const renderList = () => (
         <>
-            {(filters || actions) && (
-                <ListToolbar
-                    filters={filters}
-                    {...controllerProps}
-                    actions={actions}
-                    exporter={exporter} // deprecated, use ExporterContext instead
-                    permanentFilter={filter}
-                />
-            )}
+            {top ??
```

Comment:
the introduction of ?? forces me to do too much changes (need to upgrade typescript to >= 3.7 and eslint and prettier, so almost all files must be reformatted), so I'll revert this one.